cborg/src/Codec/CBOR/Read.hs

```diff
@@ -613,12 +629,24 @@ go_fast_end !bs (ConsumeBytes k) =
       DecodedToken sz (Fits bstr)   -> k bstr >>= go_fast_end (BS.unsafeDrop sz bs)
       DecodedToken sz (TooLong len) -> return $! SlowConsumeTokenBytes (BS.unsafeDrop sz bs) k len
 
+go_fast_end !bs (ConsumeByteArray k) =
+    case tryConsumeByteArray (BS.unsafeHead bs) bs of
+      DecodeFailure                 -> return $! SlowFail bs "expected string"
+      DecodedToken sz (Fits str)    -> k str >>= go_fast_end (BS.unsafeDrop sz bs)
+      DecodedToken sz (TooLong len) -> return $! SlowConsumeTokenByteArray (BS.unsafeDrop sz bs) k len
+
 go_fast_end !bs (ConsumeString k) =
     case tryConsumeString (BS.unsafeHead bs) bs of
       DecodeFailure                 -> return $! SlowFail bs "expected string"
       DecodedToken sz (Fits str)    -> k str >>= go_fast_end (BS.unsafeDrop sz bs)
       DecodedToken sz (TooLong len) -> return $! SlowConsumeTokenString (BS.unsafeDrop sz bs) k len
 
+go_fast_end !bs (ConsumeUtf8ByteArray k) =
+    case tryConsumeUtf8ByteArray (BS.unsafeHead bs) bs of
```

Comment:
Can't we factor this and the string/text case together too? We can make tryConsumeString be the non-copying one that returns the ByteString and then for the Text case do the conversion like `k (T.decodeUTF8 str)`.

Of course this is also a good opportunity to do the UTF8 validation properly and return failure when it fails.